### PR TITLE
N°7433 - Allow MFA configuration in the portal

### PR DESCRIPTION
**Pull request by odain-cbd**
This PR aims at customizing user profile view in portal application.

### CODE REVIEW

File: datamodels/2.x/itop-portal-base/portal/templates/layout.html.twig [hunk 1]
```diff
@@ -75,14 +75,14 @@
 		{% endif %}
 	{% endblock %}
 
+    <style>
```
Comment (Molkobain) [inline datamodels/2.x/itop-portal-base/portal/templates/layout.html.twig]: Having 1 `style` tag per extension was done on purpose. To avoid that 1 with parsing error contaminate the next ones. (Same as for the `style` node juste below)

Why did you need to move it?

Comment (eespie) [reply datamodels/2.x/itop-portal-base/portal/templates/layout.html.twig]: Right, it was a test that remained in the code.
Good catch thx Guillaume

File: datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabContentExtension.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * @copyright   Copyright (C) 2010-2024 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Portal\Hook;
+
+use Combodo\iTop\Portal\Twig\PortalTwigContext;
+
+interface iAbstractPortalTabContentExtension
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabContentExtension.php:22]: Since this interface is an extensibility point, please add the appropiate comment/docblock so that the documentation can be generated from the comment.

File: datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabExtension.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * @copyright   Copyright (C) 2010-2024 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Portal\Hook;
+
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabExtension.php:8]: Since this interface is an extensibility point, please add the appropiate comment/docblock so that the documentation can be generated from the comment.

File: datamodels/2.x/itop-portal-base/portal/src/Hook/iUserProfileTabContentExtension.php [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * @copyright   Copyright (C) 2010-2024 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Portal\Hook;
+
+interface iUserProfileTabContentExtension extends iAbstractPortalTabContentExtension
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Hook/iUserProfileTabContentExtension.php:15]: Since this interface is also an extensibility point, please add the appropiate comment/docblock so that the documentation can be generated from the comment.

File: datamodels/2.x/itop-portal-base/portal/src/Hook/iUserProfileTabExtension.php [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * @copyright   Copyright (C) 2010-2024 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Portal\Hook;
+
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Hook/iUserProfileTabExtension.php:8]: Since this interface is also an extensibility point, please add the appropiate comment/docblock so that the documentation can be generated from the comment.

File: datamodels/2.x/itop-portal-base/portal/src/Helper/ExtensibilityHelper.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * @copyright   Copyright (C) 2010-2024 Combodo SARL
+ * @license     http://opensource.org/licenses/AGPL-3.0
+ */
+
+namespace Combodo\iTop\Portal\Helper;
+
+use Combodo\iTop\Portal\Hook\iAbstractPortalTabContentExtension;
+use Combodo\iTop\Portal\Hook\iAbstractPortalTabExtension;
+use Combodo\iTop\Service\InterfaceDiscovery\InterfaceDiscovery;
+
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Helper/ExtensibilityHelper.php:12]: If I'm not mistaken, this helper will be reused when a part of the portal will provide some extensibility. Some docblock/api documentation would be welcome.

File: datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabExtension.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	 * @return float order rank
+	 */
+	public function GetTabRank(): float;
+
+	/**
+	 * Unique code for the AjaxTab
+	 *
+	 * @return string
+	 */
+	public function GetTabCode(): string;
```
Comment (dflaven) [inline datamodels/2.x/itop-portal-base/portal/src/Hook/iAbstractPortalTabExtension.php:51]: What are the allowed characters for the code? Are there any limitation ?

File: datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php [hunk 8]
```diff
@@ -106,63 +111,107 @@ public function DisplayAction(Request $oRequest, $sBrickId)
...
+		foreach ($aTabExtensions as $oPortalTabExtension) {
+			$aData['aTabsValues'][] = [
+				'code'  => $oPortalTabExtension->GetTabCode(),
+				'label' => $oPortalTabExtension->GetTabLabel(),
+			];
+		}
+
+		// Read the current tab content From iPortalTabSectionExtension
+		$aTabSectionExtensions = ExtensibilityHelper::GetInstance()->GetPortalTabContentExtensions(iUserProfileTabContentExtension::class, $sTab);
+		if (count($aTabSectionExtensions) != 0 && ! empty($_POST)){
```
Comment (steffunky) [inline datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php]: Why are you using 2 different ways to determine if an array is empty in the same condition?
Also, you need to use strict comparisons such as !==

File: datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php [hunk 9]
```diff
@@ -106,63 +111,107 @@ public function DisplayAction(Request $oRequest, $sBrickId)
...
+			$sTransactionId = utils::ReadPostedParam('transaction_id', null, utils::ENUM_SANITIZATION_FILTER_TRANSACTION_ID);
+			IssueLog::Debug(__FUNCTION__.": transaction [$sTransactionId]");
+			if (empty($sTransactionId) || !utils::IsTransactionValid($sTransactionId, false)) {
```
Comment (steffunky) [inline datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php]: Are you sure you want to check if the transaction is empty? It may be null as it is your default value but as the transaction is a float there's a non-null chance that its transaction value is 0.0 and will be thrown out by your `empty()` check.

Also, do you remove the transaction id after consuming it at some point?

File: datamodels/2.x/itop-portal-base/portal/templates/bricks/user-profile/user_info.html.twig [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+											<span class="preview">
+												<img src="{{ sUserPhotoUrl }}"/>
+											</span>
+                                        <span class="actions">
+												<span type="button" class="btn btn-default btn_edit">
+													<span class="fas fa-pencil-alt fa-fw"></span>
+													<input id="picture" type="file" name="picture" />
+												</span>
+												{#<button type="button" class="btn btn-default btn_undo" title="{{ 'UI:Button:ResetImage'|dict_s }}" disabled>
+													<span class="fas fa-undo fa-fw"></span>
```
Comment (steffunky) [inline datamodels/2.x/itop-portal-base/portal/templates/bricks/user-profile/user_info.html.twig]: Commented code can be removed

File: datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php (hunk 8 above)
Comment (odain-cbd) [reply datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php]: ok with first remark. 

!== seems useless to me when apply to count output (int returned always).

File: datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php (hunk 9 above)
Comment (odain-cbd) [reply datamodels/2.x/itop-portal-base/portal/src/Controller/UserProfileBrickController.php]: transaction_id is a string. despite incorrect PHPDOC.

transaction_id is not consumed in portal-base on purpose to let other extensions check its validity as well. that way we can display further different tab contents at the same place.
